backupccl: reset restored jobs during cluster restore

pbardea opened the pull request:
Previously, jobs were restored without modification during cluster
restore. Due to a recently discovered bug where backup may miss
non-transactional writes written to offline spans by these jobs, their
progress may no longer be accurate on the restored cluster.

IMPORT and RESTORE jobs perform non-transactional writes that may be
missed. When a cluster RESTORE brings back these OFFLINE tables, it will
also bring back its associated job. To ensure the underlying data in
these tables is correct, the jobs are now set in a reverting state so
that they can clean up after themselves.

Schema changes are another class of jobs that might perform
non-transactional writes with the backfiller. However, the revert stage
of these jobs involves another schema change. It becomes less clear what
to do when these jobs were captured in the reverting state in a backup.
So, schema change jobs are restored with their status unchanged, but
their progress field cleared. This ensures that they do not incorrectly
perceive that they've made more progress than they really have.

Release note (bug fix): Fix a bug where restored schema changes may have
assumed to have made progress that was not captured in the backup. The
restored jobs are now either canceled or reset upon cluster restore.

pbardea committed:
backupccl: reset restored jobs during cluster restore

Previously, jobs were restored without modification during cluster
restore. Due to a recently discovered bug where backup may miss
non-transactional writes written to offline spans by these jobs, their
progress may no longer be accurate on the restored cluster.

IMPORT and RESTORE jobs perform non-transactional writes that may be
missed. When a cluster RESTORE brings back these OFFLINE tables, it will
also bring back its associated job. To ensure the underlying data in
these tables is correct, the jobs are now set in a reverting state so
that they can clean up after themselves.

Schema changes are another class of jobs that might perform
non-transactional writes with the backfiller. However, the revert stage
of these jobs involves another schema change. It becomes less clear what
to do when these jobs were captured in the reverting state in a backup.
So, schema change jobs are restored with their status unchanged, but
their progress field cleared. This ensures that they do not incorrectly
perceive that they've made more progress than they really have.

Release note (bug fix): Fix a bug where restored schema changes may have
assumed to have made progress that was not captured in the backup. The
restored jobs are now either canceled or reset upon cluster restore.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/62638)
<!-- Reviewable:end -->

dt commented on pkg/ccl/backupccl/system_schema.go:
```diff
@@ -103,6 +113,120 @@ func jobsRestoreFunc(
+				}
+				payload.Details = jobspb.WrapPayloadDetails(details)
+
+				jobPayloads[jobID] = payload
```
I wonder if we should just update the job, one by one right here, rather than buffering and risking an OOM?

dt replied:
I guess details _should_ be tiny -- just the one span -- so maybe overthinking this.

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62638)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt)

___
*[pkg/ccl/backupccl/system_schema.go, line 199 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/62638#-MWgctEV9AlD5kg_YpnA-r1-199:-MY69zwHEBGA44O-XuWx:bbchzjx) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4e4645df8ea2eeccb1a2d74437bf057258833a/pkg/ccl/backupccl/system_schema.go#L199)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

I guess details _should_ be tiny -- just the one span -- so maybe I'm overthinking this.
</blockquote></details>

What I'm sad about is the number of round-trips you're going to do here. In a global cluster this could be pretty egregious. 


<!-- Sent from Reviewable.io -->

pbardea reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62638)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt)

___
*[pkg/ccl/backupccl/system_schema.go, line 199 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/62638#-MWgctEV9AlD5kg_YpnA-r1-199:-MY6EOf1-vYgDXPjdraj:b-v80ci1) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4e4645df8ea2eeccb1a2d74437bf057258833a/pkg/ccl/backupccl/system_schema.go#L199)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

What I'm sad about is the number of round-trips you're going to do here. In a global cluster this could be pretty egregious. 
</blockquote></details>

An alternative considered here was crafting a large `UPDATE ... CASE` statement that specifies the job -> payload mapping. I was a bit more concerned about a query that scales O(running jobs) though.


<!-- Sent from Reviewable.io -->

miretskiy commented:
"An alternative considered here was crafting a large UPDATE ... CASE statement
that specifies the job -> payload mapping. I was a bit more concerned about
a query that scales O(running jobs) though."

Why would you be concerned about that?  We now (hopefully) no longer have
huge payloads.  SQL is pretty reasonable as it comes
w/ ensuring we don't run out of memory; and it can plan queries in a
distributed fashion...

On Mon, Apr 12, 2021 at 2:27 PM Paul Bardea ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
>
> *Reviewable <https://reviewable.io/reviews/cockroachdb/cockroach/62638>*
> status: [image: :shipit:] complete! 0 of 0 LGTMs obtained (waiting on @dt
> <https://github.com/dt>)
>
> ------------------------------
>
> *pkg/ccl/backupccl/system_schema.go, line 199 at r1
> <https://reviewable.io/reviews/cockroachdb/cockroach/62638#-MWgctEV9AlD5kg_YpnA-r1-199:-MY6EOf1-vYgDXPjdraj:b-v80ci1>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/fa4e4645df8ea2eeccb1a2d74437bf057258833a/pkg/ccl/backupccl/system_schema.go#L199>):*
> *Previously, ajwerner wrote…*
>
> What I'm sad about is the number of round-trips you're going to do here.
> In a global cluster this could be pretty egregious.
>
> An alternative considered here was crafting a large UPDATE ... CASE
> statement that specifies the job -> payload mapping. I was a bit more
> concerned about a query that scales O(running jobs) though.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/62638#pullrequestreview-633844311>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANA4FVGA6ARPJ2UYA6S3P7DTIM3RBANCNFSM4Z2TUIGA>
> .
>

dt commented:
Yeah, putting them all in the map could OOM before we even get to SQL exec and its memory monitor. w.r.t. constructing the giant CASE, that could also easily exceed the stmt parse size or the bound param list limit. It's too bad we can't just nil it and have the SC job re-calc spans on resume if nil. Maybe something to think about when we re-work mutable progress storage.

pbardea commented:
Ideally we'd track the progress in the job's progress field rather than in the details inside the payload. It would also be convenient if ResumeSpans kept track of the work completed rather than the work that's left to do. If that was the case, the entire reset could be done in a swift `UPDATE jobstable SET progress = <bytes_of_empty_progress_proto> WHERE id IN (1,2,3)`. That would also make use of the progress column family when updating sc job progress.

dt commented on pkg/ccl/backupccl/system_schema.go:
```diff
@@ -103,6 +113,114 @@ func jobsRestoreFunc(
+					return err
+				}
+				q := fmt.Sprintf("UPDATE %s SET payload = $1 WHERE id = $2", tempTableName)
+				if _, err := executor.Exec(ctx, "updating-job-progress", txn, q, payloadBytes, jobID); err != nil {
```
I'm a little nervous this txn may never complete? all the other nodes are running pollers on the jobs table so I worry if this will just get pushed every time when trying to update a giant jobs table. maybe if we add `for update` to the select above? I'm curious what @ajwerner thinks

ajwerner replied:
SFU seems like a good idea. Especially given we really don't want the registries trying to adopt jobs anyway.

dt replied:
it occurs to me that we wrote the jobs in this txn too, so that is actually moot I guess?

ajwerner replied:
Turns out nobody is reading this

pbardea replied:
I did want to check what the expected semantics are here wrt the txn use between the select above and this update given the new streaming internal executor. Is it safe to update this data while iterating over the streamed results?

ajwerner commented on pkg/ccl/backupccl/system_schema.go:
```diff
@@ -103,6 +113,114 @@ func jobsRestoreFunc(
+				}
+				if len(payload.DescriptorIDs) != 1 {
+					continue
+				}
```
if the mutation ID is 0 you don't need to do anything.

pbardea replied:
Done.

ajwerner commented on pkg/ccl/backupccl/system_schema.go:
```diff
@@ -103,6 +113,114 @@ func jobsRestoreFunc(
+	)
+
+	err := func() (err error) {
+		// makeJobsList writes the given jobs as a SQL tuple set to the
```
Is there something that has prevented these jobs from being adopted?

pbardea replied:
This updating is being done in a temporary table before they're moved over to `system.jobs`.

ajwerner replied:
In that case, we could batch this in separate transactions if we were worried about the transaction being too big. It doesn't seem that we're super concerned about that. Is there a strong need for atomicity here updating the jobs table and other things happening in this txn?

pbardea replied:
> Is there a strong need for atomicity here updating the jobs table and other things happening in this txn?

This txn updates the job state to indicate that the system tables have been restored (`details.SystemTablesRestored`). This information is used to protect against the case where:
- The job migrates the temporary system table data
- The job restarts and retries to re-restore the temporary system table's data, but on restore AddSSTable will now shadow some keys with a different value and throws an error.

I don't think that there is any restriction that requires this update to be in the same transaction as the updating the `system.jobs` table since we don't allow stale reads.

That being said, this change can be simplified by splitting the system table restoration step into 2 steps:
1. Updating/migrating the system data in the staging table. Even if this phase is idempotent, we cannot re-attempt to restore data over this migrated data, so we need to persist state in the job keeping track of migrations.
2. Actually moving the data over to the live system tables. Previously, this may have not been idempotent and we used the previously mentioned state on the job to check if we've restored this system table, but I now realize that this can be made to be idempotent by upserting the values, rather than inserting them. Perhaps enforcing that these are idempotent is the way to go.

ajwerner replied:
> but I now realize that this can be made to be idempotent by upserting the values, rather than inserting them. Perhaps enforcing that these are idempotent is the way to go.

Rather than `UPSERT` it feels like what you want might be `INSERT ... ON CONFLICT DO NOTHING`. That way if the job has already been adopted we don't muck with it.

pbardea replied:
Did a bit of the refactoring mentioned here. If I add batching, I think that could help with the number of roundtrips. Will tackle that next.

ajwerner commented on pkg/ccl/backupccl/system_schema.go:
```diff
@@ -103,6 +113,114 @@ func jobsRestoreFunc(
+					continue
+				}
+				allSpans := table.PrimaryIndexSpan(execCfg.Codec)
+				for i := range details.ResumeSpanList {
```
Can we make this idempotent and avoid the update if the resume spans are correct?

pbardea replied:
Done.

pbardea committed:
backupccl: make job system table restoration idempotent

Previously, custom implementations of restoring system tables during
cluster restore may have not been idempotent. As such, a map was used to
track when particular system tables had been restored. This was fragile.
This change updates the system table restoration logic to be idempotent
for all custom implementation (only the jobs table needed updating).

Release note: None
